types.d.ts

```diff
@@ -0,0 +1,1453 @@
+     */
+    getOperationForVault(vault: string): VaultOperation | null;
+
+    /**
+     * Register a read miss
+     *
+     * @param {string} vault
+     *
+     * @memberof VaultValue
+     */
+    registerReadMiss(vault: string): void;
+
+    /**
+     * Check if the value previously did not exist
+     *
+     * @returns {boolean} true if the value previously did not exist
+     *
+     * @memberof VaultValue
+     */
+    didNotExist(): boolean;
+
+    /**
+     * Schedule a value add to the different vault backends
+     *
+     * @param {string} mediaType
+     * @param {*} data
+     * @param {string} encoding
+     *
+     * @memberof VaultValue
+     */
+    add(mediaType: string, data: any, encoding: string): void;
+
+    /**
+     * Schedule a data set on the different vault backends
+     *
+     * @param {string} mediaType
+     * @param {*} data
+     * @param {string} encoding
+     *
+     * @memberof VaultValue
+     */
+    set(mediaType: string, data: any, encoding: string): void;
+
+    /**
+     * Schedule the set of the expiration time for the vault value (on supported vault backends)
+     *
+     * @param {number} expirationTime
+     *
+     * @memberof VaultValue
+     */
+    touch(expirationTime: number): void;
+
+    /**
+     * Mark the vault value for deletion in the different vault backends
+     *
+     * @memberof VaultValue
+     */
+    del(): void;
+
+    /**
+     * Retrieve the data diff for this vault value.
+     *
+     * This will only works on values with a mediaType supporting diffs (so currently, only tomes)
+     *
+     * @returns {(object[]|null)}
+     *
+     * @memberof VaultValue
+     */
+    getDiff(): object[]|null;
+
+    /**
+     * Apply a diff to the vault value.
+     *
+     * This will only works on values with a mediaType supporting diffs (so currently, only tomes)
+     *
+     * @param {*} diff
+     *
+     * @memberof VaultValue
+     */
+    applyDiff(diff: any): void;
+}
+
+declare class Archivist {
+    /**
+     * Check whether a given exists in any of our vaults
+     *
+     * @param {string} topicName
+     * @param {ArchivistIndex} index
+     * @param {ArchivistExistsCallback} callback
+     *
+     * @memberOf Archivist
+     */
+    exists(topicName: string, index: ArchivistIndex, callback: ArchivistExistsCallback): void;
+
+    /**
+     * Retrieve a value
+     *
+     * This method is a wrapper around `getValue` which directly returns ArchivistValue.value;
+     * in most cases, you will want to call `get` instead to get the actual data.
+     *
+     * @param {string} topicName
+     * @param {ArchivistIndex} index
+     * @param {ArchivistOptions} options
+     * @param {Callback} callback
+     *
+     * @memberOf Archivist
+     */
+    get(topicName: string, index: ArchivistIndex, options: ArchivistGetOptions, callback: ArchivistGetCallback): void;
+    get(topicName: string, index: ArchivistIndex, callback: ArchivistGetCallback): void;
```

Comment:
It would be nice to have:

```typescript
type TypedArchivistGetCallback<T> = (err: Error, val: T) => void;
get<T>(topicName: string, index: ArchivistIndex, callback: TypedArchivistGetCallback<T>): void;
```

So that in my code I can do:

```typescript
state.archivist.get<IGame>("game", { gameId: "gameId" }, (err, val) => {
  // val is an IGame instance there
});
```